File: wear/src/main/java/com/pdyjak/powerampwear/player/PlayerViewModel.java
package com.pdyjak.powerampwear.player;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.android.gms.wearable.MessageEvent;
import com.maxmpz.poweramp.player.PowerampAPI;
import com.pdyjak.powerampwear.MessageExchangeHelper;
import com.pdyjak.powerampwear.MessageListener;
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwear.music_browser.albums.AlbumItem;
import com.pdyjak.powerampwear.music_browser.categories.CategoryItem;
import com.pdyjak.powerampwear.music_browser.folders.FolderItem;
import com.pdyjak.powerampwear.settings.SettingsManager;
import com.pdyjak.powerampwearcommon.events.PlayingModeChangedEvent;
import com.pdyjak.powerampwearcommon.events.StatusChangedEvent;
import com.pdyjak.powerampwearcommon.events.TrackChangedEvent;
import com.pdyjak.powerampwearcommon.events.TrackPositionSyncEvent;
import com.pdyjak.powerampwearcommon.requests.FindParentRequest;
import com.pdyjak.powerampwearcommon.requests.GetFilesRequest;
import com.pdyjak.powerampwearcommon.requests.RequestsPaths;
import com.pdyjak.powerampwearcommon.responses.FindParentResponse;
import com.pdyjak.powerampwearcommon.responses.Parent;

import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.WeakHashMap;

class PlayerViewModel implements MessageListener {
    private static final int DEFAULT_TIMEOUT = 5000;
    private static final int TRACK_POSITION_UPDATE_DELAY = 1000;
    private static final int SECONDS_PER_TICK = TRACK_POSITION_UPDATE_DELAY / 1000;

    private static final int REPEAT_MAX_LEVEL = 3;
    private static final int SHUFFLE_MAX_LEVEL = 4;


    interface UiElementsVisibilityListener {
        void onClockVisibilityChanged();
        void onProgressbarVisibilityChanged();
    }

    interface CommonEventsListener {
        void onStateChanged();
        void onPauseChanged();
        void onTrackInfoChanged();
    }

    interface RepeatShuffleModesListener {
        void onRepeatModeChanged();
        void onShuffleModeChanged();
    }

    interface TrackPositionListener {
        void onPositionChanged(int position, int duration); // both in seconds
    }

    enum State {
        Loading,
        Failure,
        Ok
    }

    enum ShuffleMode {
        Off(0),
        All(1),
        Songs(2),
        Lists(3),
        SongsLists(4);

        final int value;
        ShuffleMode(int value) {
            this.value = value;
        }

        @Nullable
        static ShuffleMode from(int value) {
            ShuffleMode modes[] = ShuffleMode.values();
            if (value < modes.length) return modes[value];
            return null;
        }
    }

    enum RepeatMode {
        Off(0),
        List(1),
        AdvanceList(2),
        Song(3);

        final int value;
        RepeatMode(int value) {
            this.value = value;
        }

        @Nullable
        static RepeatMode from(int value) {
            RepeatMode modes[] = RepeatMode.values();
            if (value < modes.length) return modes[value];
            return null;
        }
    }

    private class SettingsListener extends SettingsManager.Listener {
        @Override
        public void onClockSettingChanged() {
            refreshClockVisibilityStatus();
        }
    }

    private class AmbientModeListener implements AmbientModeStateProvider.Listener {
        @Override
        public void onAmbientModeStateChanged() {
            if (mAmbientModeStateProvider.isInAmbientMode()) {
                mHandler.removeCallbacks(mTrackTimeUpdateRunnable);
            } else {
                mMessageExchangeHelper.sendRequest(RequestsPaths.SYNC_TRACK_POSITION);
            }
            refreshProgressbarVisibilityStatus();
        }
    }

    @NonNull
    private final SettingsManager mSettingsManager;
    @NonNull
    private final MessageExchangeHelper mMessageExchangeHelper;
    @NonNull
    private final AmbientModeStateProvider mAmbientModeStateProvider;
    @NonNull
    private final MusicLibraryNavigator mMusicLibraryNavigator;
    @NonNull
    private final SettingsListener mClockSettingListener = new SettingsListener();
    @NonNull
    private final AmbientModeListener mAmbientModeListener = new AmbientModeListener();
    @NonNull
    private final Set<CommonEventsListener> mListeners =
            Collections.newSetFromMap(new WeakHashMap<CommonEventsListener, Boolean>());
    @NonNull
    private final Set<RepeatShuffleModesListener> mRepeatShuffleListeners =
            Collections.newSetFromMap(new WeakHashMap<RepeatShuffleModesListener, Boolean>());
    @NonNull
    private final Set<UiElementsVisibilityListener> mUiElementsVisibilityListeners =
            Collections.newSetFromMap(new WeakHashMap<UiElementsVisibilityListener, Boolean>());
    @NonNull
    private final Handler mHandler = new Handler(Looper.getMainLooper());
    @NonNull
    private final Runnable mTimeoutRunnable = new Runnable() {
        @Override
        public void run() {
            setState(State.Failure);
        }
    };
    @NonNull
    private final Runnable mTrackTimeUpdateRunnable = new Runnable() {
        @Override
        public void run() {
            mCurrentTrackPosition += SECONDS_PER_TICK;
            notifyTrackPositionChanged();
            mHandler.removeCallbacks(mTrackTimeUpdateRunnable);
            mHandler.postDelayed(mTrackTimeUpdateRunnable, TRACK_POSITION_UPDATE_DELAY);
        }
    };
    @Nullable
    private WeakReference<TrackPositionListener> mTrackPositionListener;

    private boolean mTimeoutRunnablePosted;
    @NonNull
    private State mCurrentState = State.Loading;
    @NonNull
    private RepeatMode mRepeatMode = RepeatMode.Off;
    @NonNull
    private ShuffleMode mShuffleMode = ShuffleMode.Off;
    @Nullable
    private String mTitle;
    @Nullable
    private String mArtistAlbum;
    private boolean mPaused = true;
    private boolean mShouldShowClock;
    private boolean mInAmbientMode;
    private int mCurrentTrackDuration;
    private int mCurrentTrackPosition;

    PlayerViewModel(@NonNull SettingsManager settingsManager,
                    @NonNull MessageExchangeHelper helper,
                    @NonNull AmbientModeStateProvider ambientModeStateProvider,
                    @NonNull MusicLibraryNavigator musicLibraryNavigator) {
        mSettingsManager = settingsManager;
        mMessageExchangeHelper = helper;
        mAmbientModeStateProvider = ambientModeStateProvider;
        mMusicLibraryNavigator = musicLibraryNavigator;
        mShouldShowClock = mSettingsManager.shouldShowClock();
        mInAmbientMode = mAmbientModeStateProvider.isInAmbientMode();
    }

    void addListenerWeakly(@NonNull CommonEventsListener listener) {
        mListeners.add(listener);
    }

    void addListenerWeakly(@NonNull RepeatShuffleModesListener listener) {
        mRepeatShuffleListeners.add(listener);
    }

    void addListenerWeakly(@NonNull UiElementsVisibilityListener listener) {
        mUiElementsVisibilityListeners.add(listener);
    }

    void removeListener(@NonNull CommonEventsListener listener) {
        mListeners.remove(listener);
    }

    void removeListener(@NonNull RepeatShuffleModesListener listener) {
        mRepeatShuffleListeners.remove(listener);
    }

    void removeListener(@NonNull UiElementsVisibilityListener listener) {
        mUiElementsVisibilityListeners.remove(listener);
    }

    void setTrackPositionListenerWeakly(@Nullable TrackPositionListener listener) {
        if (listener == null) {
            mTrackPositionListener = null;
            return;
        }
        mTrackPositionListener = new WeakReference<>(listener);
    }

    void onResume() {
        if (!mTimeoutRunnablePosted) {
            mTimeoutRunnablePosted = true;
            mHandler.postDelayed(mTimeoutRunnable, DEFAULT_TIMEOUT);
        }
        mMessageExchangeHelper.addMessageListenerWeakly(this);
        mMessageExchangeHelper.sendRequest(RequestsPaths.REFRESH_TRACK_INFO);
        mSettingsManager.addSettingsListener(mClockSettingListener);
        mAmbientModeStateProvider.addAmbientModeListener(mAmbientModeListener);
        refreshClockVisibilityStatus();
        refreshProgressbarVisibilityStatus();
    }

    private void refreshClockVisibilityStatus() {
        boolean showClock = mSettingsManager.shouldShowClock();
        if (showClock == mShouldShowClock) return;
        mShouldShowClock = showClock;
        Set<UiElementsVisibilityListener> copy = new HashSet<>(mUiElementsVisibilityListeners);
        for (UiElementsVisibilityListener listener : copy) {
            listener.onClockVisibilityChanged();
        }
    }

    private void refreshProgressbarVisibilityStatus() {
        boolean ambient = mAmbientModeStateProvider.isInAmbientMode();
        if (mInAmbientMode == ambient) return;
        mInAmbientMode = ambient;
        Set<UiElementsVisibilityListener> copy = new HashSet<>(mUiElementsVisibilityListeners);
        for (UiElementsVisibilityListener listener : copy) {
            listener.onProgressbarVisibilityChanged();
        }
    }

    void onPause() {
        mMessageExchangeHelper.removeMessageListener(this);
        mSettingsManager.removeListener(mClockSettingListener);
        mAmbientModeStateProvider.removeAmbientModeListener(mAmbientModeListener);
        removeTimeoutCallbackFromHandler();
        mHandler.removeCallbacks(mTrackTimeUpdateRunnable);
    }

    @NonNull
    State getCurrentState() {
        return mCurrentState;
    }

    @NonNull
    RepeatMode getRepeatMode() {
        return mRepeatMode;
    }

    @NonNull
    ShuffleMode getShuffleMode() {
        return mShuffleMode;
    }

    @Nullable
    String getTitle() {
        return mTitle;
    }

    @Nullable
    String getArtistAlbum() {
        return mArtistAlbum;
    }

    boolean isPaused() {
        return mPaused;
    }

    boolean shouldShowClock() {
        return mShouldShowClock;
    }

    boolean shouldShowProgressbar() {
        return !mInAmbientMode;
    }

    private void removeTimeoutCallbackFromHandler() {
        if (mTimeoutRunnablePosted) {
            mTimeoutRunnablePosted = false;
            mHandler.removeCallbacks(mTimeoutRunnable);
        }
    }

    @Override
    public void onMessageReceived(@NonNull MessageEvent messageEvent) {
        switch (messageEvent.getPath()) {
            case TrackChangedEvent.PATH:
                removeTimeoutCallbackFromHandler();
                setState(State.Ok);
                processTrackInfo(TrackChangedEvent.fromBytes(messageEvent.getData()));
                break;

            case StatusChangedEvent.PATH:
                processStatusInfo(StatusChangedEvent.fromBytes(messageEvent.getData()));
                break;

            case PlayingModeChangedEvent.PATH:
                processPlayingModeInfo(PlayingModeChangedEvent.fromBytes(messageEvent.getData()));
                break;

            case TrackPositionSyncEvent.PATH:
                processTrackPositionSyncInfo(TrackPositionSyncEvent.fromBytes(
                        messageEvent.getData()));
                break;

            case FindParentResponse.PATH:
                goToLibrary(FindParentResponse.fromBytes(messageEvent.getData()));
                break;
        }
    }


    private void processTrackPositionSyncInfo(
            @NonNull TrackPositionSyncEvent trackPositionSyncEvent) {
        mCurrentTrackPosition = trackPositionSyncEvent.position;
        notifyTrackPositionChanged();
        if (!mPaused) {
            mHandler.removeCallbacks(mTrackTimeUpdateRunnable);
            mHandler.postDelayed(mTrackTimeUpdateRunnable, TRACK_POSITION_UPDATE_DELAY);
        }
    }

    private void notifyTrackPositionChanged() {
        TrackPositionListener listener = mTrackPositionListener == null
                ? null : mTrackPositionListener.get();
        if (listener == null || mCurrentTrackPosition > mCurrentTrackDuration) return;
        listener.onPositionChanged(mCurrentTrackPosition, mCurrentTrackDuration);
    }

    private void processPlayingModeInfo(@NonNull PlayingModeChangedEvent playingModeChangedEvent) {
        ShuffleMode shuffleMode = ShuffleMode.from(playingModeChangedEvent.shuffleMode);
        if (shuffleMode != null) setShuffleMode(shuffleMode);
        RepeatMode repeatMode = RepeatMode.from(playingModeChangedEvent.repeatMode);
        if (repeatMode != null) setRepeatMode(repeatMode);
    }

    private void processTrackInfo(@NonNull TrackChangedEvent event) {
        boolean changed = setTitle(event.title);
        mCurrentTrackDuration = event.duration;
        String secondLine = "";
        if (event.artist != null) {
            secondLine = event.artist;
        }
        if (event.album != null) {
            if (secondLine.length() > 0) secondLine += " - ";
            secondLine += event.album;
        }
        changed |= setArtistAlbum(secondLine);
        if (changed) {
            Set<CommonEventsListener> copy = new HashSet<>(mListeners);
            for (CommonEventsListener listener : copy) {
                listener.onTrackInfoChanged();
            }
        }
    }

    private void processStatusInfo(@NonNull StatusChangedEvent event) {
        setPaused(event.status != PowerampAPI.Status.TRACK_PLAYING || event.paused);
    }

    void togglePlayPause() {
        setPaused(!mPaused);
        if (mPaused) {
            mHandler.removeCallbacks(mTrackTimeUpdateRunnable);
        } else {
            mMessageExchangeHelper.sendRequest(RequestsPaths.SYNC_TRACK_POSITION);
        }
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_PLAY_PAUSE);
    }

    void toggleRepeatMode() {
        RepeatMode next = mRepeatMode.value == REPEAT_MAX_LEVEL ? RepeatMode.from(0)
                : RepeatMode.from(mRepeatMode.value + 1);
        if (next == null) return;
        setRepeatMode(next);
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_REPEAT_MODE);
    }

    void toggleShuffleMode() {
        ShuffleMode next = mShuffleMode.value == SHUFFLE_MAX_LEVEL ? ShuffleMode.from(0)
                : ShuffleMode.from(mShuffleMode.value + 1);
        if (next == null) return;
        setShuffleMode(next);
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_SHUFFLE_MODE);
    }

    void previousTrack() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.PREV_TRACK);
    }

    void nextTrack() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.NEXT_TRACK);
    }

    void volumeDown() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.VOLUME_DOWN);
    }

    void volumeUp() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.VOLUME_UP);
    }

    void goToLibrary() {
        FindParentRequest request = new FindParentRequest(-1);
        mMessageExchangeHelper.sendRequest(FindParentRequest.PATH, request);
    }

    private void goToLibrary(@NonNull FindParentResponse response) {
        Parent parent = response.parent;
        if (parent == null) {
            // All tracks
            CategoryItem categoryItem = new CategoryItem(mMusicLibraryNavigator,
                    GetFilesRequest.PATH, 0, 0);
            mMusicLibraryNavigator.selectCategory(categoryItem, true, response.title);
            return;
        }

        switch (parent.type) {
            case Queue:
            case Folder:
                FolderItem folderItem = new FolderItem(mMusicLibraryNavigator, parent.id, null,
                        null);
                mMusicLibraryNavigator.selectFolder(folderItem, true, response.title);
                break;

            case Album:
                AlbumItem albumItem = new AlbumItem(mMusicLibraryNavigator, parent.id, null, null);
                mMusicLibraryNavigator.selectAlbum(albumItem, true, response.title);
                break;
        }
    }

    private void setState(@NonNull State state) {
        if (mCurrentState == state) return;
        mCurrentState = state;
        Set<CommonEventsListener> copy = new HashSet<>(mListeners);
        for (CommonEventsListener listener : copy) {
            listener.onStateChanged();
        }
    }

    private void setPaused(boolean paused) {
        if (mPaused == paused) return;
        mPaused = paused;
        Set<CommonEventsListener> copy = new HashSet<>(mListeners);
        for (CommonEventsListener listener : copy) {
            listener.onPauseChanged();
        }
    }

    private void setRepeatMode(@NonNull RepeatMode mode) {
        if (mRepeatMode == mode) return;
        mRepeatMode = mode;
        Set<RepeatShuffleModesListener> copy = new HashSet<>(mRepeatShuffleListeners);
        for (RepeatShuffleModesListener listener : copy) {
            listener.onRepeatModeChanged();
        }
    }

    private void setShuffleMode(@NonNull ShuffleMode mode) {
        if (mShuffleMode == mode) return;
        mShuffleMode = mode;
        Set<RepeatShuffleModesListener> copy = new HashSet<>(mRepeatShuffleListeners);
        for (RepeatShuffleModesListener listener : copy) {
            listener.onShuffleModeChanged();
        }
    }

    private boolean setTitle(@Nullable String title) {
        if (equals(title, mTitle)) return false;
        mTitle = title;
        return true;
    }

    private boolean setArtistAlbum(@NonNull String artistAlbum) {
        if (artistAlbum.equals(mArtistAlbum)) return false;
        mArtistAlbum = artistAlbum;
        return true;
    }

    private static boolean equals(@Nullable Object o1, @Nullable Object o2) {
        return o1 == null ? o2 == null : o1.equals(o2);
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/player/PlayerViewModel.kt
package com.pdyjak.powerampwear.player

import android.os.Handler
import android.os.Looper
import com.google.android.gms.wearable.MessageEvent
import com.maxmpz.poweramp.player.PowerampAPI
import com.pdyjak.powerampwear.MessageExchangeHelper
import com.pdyjak.powerampwear.MessageListener
import com.pdyjak.powerampwear.common.Event
import com.pdyjak.powerampwear.common.SimpleEvent
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator
import com.pdyjak.powerampwear.music_browser.albums.AlbumItem
import com.pdyjak.powerampwear.music_browser.categories.CategoryItem
import com.pdyjak.powerampwear.music_browser.folders.FolderItem
import com.pdyjak.powerampwear.settings.SettingsManager
import com.pdyjak.powerampwearcommon.events.PlayingModeChangedEvent
import com.pdyjak.powerampwearcommon.events.StatusChangedEvent
import com.pdyjak.powerampwearcommon.events.TrackChangedEvent
import com.pdyjak.powerampwearcommon.events.TrackPositionSyncEvent
import com.pdyjak.powerampwearcommon.requests.FindParentRequest
import com.pdyjak.powerampwearcommon.requests.GetFilesRequest
import com.pdyjak.powerampwearcommon.requests.RequestsPaths
import com.pdyjak.powerampwearcommon.responses.FindParentResponse
import com.pdyjak.powerampwearcommon.responses.Parent

data class PositionChangedEventArgs(val position: Int, val duration: Int)

class PlayerViewModel(private val mSettingsManager: SettingsManager,
                      private val mMessageExchangeHelper: MessageExchangeHelper,
                      private val mAmbientModeStateProvider: AmbientModeStateProvider,
                      private val mMusicLibraryNavigator: MusicLibraryNavigator)
    : MessageListener {

    private companion object {
        const val DEFAULT_TIMEOUT = 5000
        const val TRACK_POSITION_UPDATE_DELAY = 1000
        const val SECONDS_PER_TICK = TRACK_POSITION_UPDATE_DELAY / 1000

        const val REPEAT_MAX_LEVEL = 3
        const val SHUFFLE_MAX_LEVEL = 4
    }

    enum class State {
        Loading,
        Failure,
        Ok
    }

    enum class ShuffleMode(val value: Int) {
        Off(0),
        All(1),
        Songs(2),
        Lists(3),
        SongsLists(4);

        companion object {
            fun from(ordinal: Int): ShuffleMode? = values().let {
                return if (it.size < ordinal) it[ordinal] else null
            }
        }
    }

    enum class RepeatMode(val value: Int) {
        Off(0),
        List(1),
        AdvanceList(2),
        Song(3);

        companion object {
            fun from(ordinal: Int): RepeatMode? = values().let {
                return if (it.size < ordinal) it[ordinal] else null
            }
        }
    }

    val onClockVisibilityChanged = SimpleEvent()
    val onProgressbarVisibilityChanged = SimpleEvent()
    val onStateChanged = SimpleEvent()
    val onPauseChanged = SimpleEvent()
    val onTrackInfoChanged = SimpleEvent()
    val onRepeatModeChanged = SimpleEvent()
    val onShuffleModeChanged = SimpleEvent()
    val onPositionChanged = Event<PositionChangedEventArgs>()

    private var mCurrentTrackPosition = 0
    private var mCurrentTrackDuration = 0

    private val mClockSettingsHandler = { shouldShowClock = mSettingsManager.shouldShowClock() }
    private val mAmbientModeHandler = {
        if (mAmbientModeStateProvider.isInAmbientMode) {
            mHandler.removeCallbacks(mTrackTimeUpdateRunnable)
        } else {
            mMessageExchangeHelper.sendRequest(RequestsPaths.SYNC_TRACK_POSITION)
        }
        shouldShowProgressbar = !mAmbientModeStateProvider.isInAmbientMode
    }

    private val mTimeoutRunnable = Runnable { state = State.Failure }

    private val mTrackTimeUpdateRunnable = object: Runnable {
        override fun run() {
            mCurrentTrackPosition += SECONDS_PER_TICK
            notifyTrackPositionChanged()
            mHandler.removeCallbacks(this)
            mHandler.postDelayed(this, TRACK_POSITION_UPDATE_DELAY.toLong())
        }
    }

    private val mHandler = Handler(Looper.getMainLooper())
    private var mActive = false
    private var mTimeoutRunnablePosted = false

    var state: State = State.Loading
        get
        private set(value) {
            if (field == value) return
            field = value
            onStateChanged()
        }

    var paused: Boolean = true
        get
        private set(value) {
            if (field == value) return
            field = value
            onPauseChanged()
        }

    var repeatMode: RepeatMode = RepeatMode.Off
        get
        private set(value) {
            if (field == value) return
            field = value
            onRepeatModeChanged()
        }

    var shuffleMode: ShuffleMode = ShuffleMode.Off
        get
        private set(value) {
            if (field == value) return
            field = value
            onShuffleModeChanged()
        }

    var shouldShowClock: Boolean = mSettingsManager.shouldShowClock()
        get
        private set(value) {
            if (field == value) return
            field = value
            onClockVisibilityChanged()
        }

    var shouldShowProgressbar: Boolean = !mAmbientModeStateProvider.isInAmbientMode
        get
        private set(value) {
            if (field == value) return
            field = value
            onProgressbarVisibilityChanged()
        }

    var title: String = ""
        get
        private set

    var artistAlbum: String = ""
        get
        private set

    fun onResume() {
        if (mActive) return
        mActive = true
        if (!mTimeoutRunnablePosted) {
            mTimeoutRunnablePosted = true
            mHandler.postDelayed(mTimeoutRunnable, DEFAULT_TIMEOUT.toLong())
        }
        mMessageExchangeHelper.addMessageListenerWeakly(this)
        mMessageExchangeHelper.sendRequest(RequestsPaths.REFRESH_TRACK_INFO)
        mSettingsManager.onClockSettingChanged += mClockSettingsHandler
        mAmbientModeStateProvider.onAmbientModeChanged += mAmbientModeHandler
        shouldShowClock = mSettingsManager.shouldShowClock()
        shouldShowProgressbar = !mAmbientModeStateProvider.isInAmbientMode
    }

    fun onPause() {
        if (!mActive) return
        mActive = false
        mMessageExchangeHelper.removeMessageListener(this)
        mSettingsManager.onClockSettingChanged -= mClockSettingsHandler
        mAmbientModeStateProvider.onAmbientModeChanged -= mAmbientModeHandler
        removeTimeoutCallbackFromHandler()
        mHandler.removeCallbacks(mTrackTimeUpdateRunnable)
    }

    private fun removeTimeoutCallbackFromHandler() {
        if (!mTimeoutRunnablePosted) return
        mTimeoutRunnablePosted = false
        mHandler.removeCallbacks(mTimeoutRunnable)
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        when (messageEvent.path) {
            TrackChangedEvent.PATH -> {
                removeTimeoutCallbackFromHandler()
                state = State.Ok
                processTrackInfo(TrackChangedEvent.fromBytes(messageEvent.data))
            }

            StatusChangedEvent.PATH -> {
                processStatusInfo(StatusChangedEvent.fromBytes(messageEvent.data))
            }

            PlayingModeChangedEvent.PATH -> {
                processPlayingModeInfo(PlayingModeChangedEvent.fromBytes(messageEvent.data))
            }

            TrackPositionSyncEvent.PATH -> {
                processTrackPositionSyncInfo(TrackPositionSyncEvent.fromBytes(messageEvent.data))
            }

            FindParentResponse.PATH -> {
                goToLibrary(FindParentResponse.fromBytes(messageEvent.data))
            }
        }
    }

    private fun processTrackInfo(event: TrackChangedEvent) {
        val changed = title != event.title
        title = event.title
        mCurrentTrackDuration = event.duration
        var artistAlbum = ""
        if (event.artist != null) artistAlbum = event.artist
        if (event.album != null) {
            if (artistAlbum.isNotEmpty()) artistAlbum += " - "
            artistAlbum += event.album
        }
        this.artistAlbum = artistAlbum
        if (changed || this.artistAlbum != artistAlbum) onTrackInfoChanged()
    }

    private fun processStatusInfo(event: StatusChangedEvent) {
        paused = event.status != PowerampAPI.Status.TRACK_PLAYING || event.paused
    }

    private fun processPlayingModeInfo(event: PlayingModeChangedEvent) {
        val shuffle = ShuffleMode.from(event.shuffleMode)
        if (shuffle != null) shuffleMode = shuffle
        val repeat = RepeatMode.from(event.repeatMode)
        if (repeat != null) repeatMode = repeat
    }

    private fun processTrackPositionSyncInfo(event: TrackPositionSyncEvent) {
        mCurrentTrackPosition = event.position
        notifyTrackPositionChanged()
        if (paused) return
        mHandler.removeCallbacks(mTrackTimeUpdateRunnable)
        mHandler.postDelayed(mTrackTimeUpdateRunnable, TRACK_POSITION_UPDATE_DELAY.toLong())
    }

    private fun notifyTrackPositionChanged() {
        onPositionChanged(PositionChangedEventArgs(mCurrentTrackPosition, mCurrentTrackDuration))
    }

    private fun goToLibrary(response: FindParentResponse) {
        val parent = response.parent
        if (parent == null) {
            // All tracks
            val categoryItem = CategoryItem(mMusicLibraryNavigator, GetFilesRequest.PATH, 0, 0)
            mMusicLibraryNavigator.selectCategory(categoryItem, true, response.title)
            return
        }

        if (parent.type == Parent.Type.Queue || parent.type == Parent.Type.Folder) {
            val folderItem = FolderItem(mMusicLibraryNavigator, parent.id, null, null)
            mMusicLibraryNavigator.selectFolder(folderItem, true, response.title)
        } else if (parent.type == Parent.Type.Album) {
            val albumItem = AlbumItem(mMusicLibraryNavigator, parent.id, null, null)
            mMusicLibraryNavigator.selectAlbum(albumItem, true, response.title)
        }
    }

    fun goToLibrary() {
        mMessageExchangeHelper.sendRequest(FindParentRequest.PATH,
                FindParentRequest(-1))
    }

    fun togglePlayPause() {
        paused = !paused
        if (paused) {
            mHandler.removeCallbacks(mTrackTimeUpdateRunnable)
        } else {
            mMessageExchangeHelper.sendRequest(RequestsPaths.SYNC_TRACK_POSITION)
        }
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_PLAY_PAUSE)
    }

    fun toggleRepeatMode() {
        val next = (if (repeatMode.value == REPEAT_MAX_LEVEL)
            RepeatMode.from(0)
        else
            RepeatMode.from(repeatMode.value + 1)) ?: return
        repeatMode = next
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_REPEAT_MODE)
    }

    fun toggleShuffleMode() {
        val next = (if (shuffleMode.value == SHUFFLE_MAX_LEVEL)
            ShuffleMode.from(0)
        else
            ShuffleMode.from(shuffleMode.value + 1)) ?: return
        shuffleMode = next
        mMessageExchangeHelper.sendRequest(RequestsPaths.TOGGLE_SHUFFLE_MODE)
    }

    fun previousTrack() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.PREV_TRACK)
    }

    fun nextTrack() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.NEXT_TRACK)
    }

    fun volumeDown() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.VOLUME_DOWN)
    }

    fun volumeUp() {
        mMessageExchangeHelper.sendRequest(RequestsPaths.VOLUME_UP)
    }
}
